echo -e "run\tuser\thost\tVirussequence\tsample\ttotalreads\treadshostR1\treadshost\t%readshost\treadsvirus\t%readsvirus\tunmappedreads\t%unmapedreads\tmedianDPcoveragevirus\tCoverage>10x(%)\tVariantsinconsensusx10\tMissenseVariants\t%Ns10x\tLineage\tanalysis_date" > mapping_illumina_$(date '+%Y%m%d').tab
RUN=$(ls -l ../../RAW/ | cut -d'/' -f4 | sort -u | grep -v 'total' | head -n1)
USER=$(pwd | cut -d '/' -f6 | cut -d '_' -f4)
HOST=$(pwd | cut -d '/' -f8 | cut -d '_' -f4 | tr '[:upper:]' '[:lower:]' | sed 's/.*/\u&/')

cat samples_ref.txt | while read in
do
        arr=($in);
	echo -e "${RUN}\t${USER}\t${HOST}\t${arr[1]}\t${arr[0]}\t$(grep 'total_reads' ${arr[1]}*/fastp/${arr[0]}.fastp.json | head -n2 | tail -n1 | cut -d ':' -f2 | sed 's/,//g')\t$(cat ${arr[1]}*/kraken2/${arr[0]}.kraken2.report.txt | grep -v 'unclassified' | cut -f3 | awk '{s+=$1}END{print s}')\t$(cat ${arr[1]}*/kraken2/${arr[0]}.kraken2.report.txt | grep -v 'unclassified' | cut -f3 | awk '{s+=$1}END{print s}' | awk '{print ($1*2)}')\t$(echo "$(echo -e "$(cat ${arr[1]}*/kraken2/${arr[0]}.kraken2.report.txt| grep -v 'unclassified' | cut -f3 | awk '{s+=$1}END{print s}' | awk '{print ($1*2)}')\t$(grep 'total_reads' ${arr[1]}*/fastp/${arr[0]}.fastp.json | head -n2 | tail -n1 | cut -d ':' -f2 | sed 's/,//g')" | awk '{print (($1*100)/$2)}')")\t$(cat ${arr[1]}*/variants/bowtie2/samtools_stats/${arr[0]}.sorted.bam.flagstat | grep '+ 0 mapped' | cut -d ' ' -f1)\t$(cat ${arr[1]}*/multiqc/summary_variants_metrics_mqc.csv | grep ^${arr[0]}, | cut -d ',' -f5)\t$(echo "$(echo -e "$(grep 'total_reads' ${arr[1]}*/fastp/${arr[0]}.fastp.json | head -n2 | tail -n1 | cut -d ':' -f2 | sed 's/,//g')\t$(cat ${arr[1]}*/kraken2/${arr[0]}.kraken2.report.txt | grep -v 'unclassified' | cut -f3 | awk '{s+=$1}END{print s}' | awk '{print ($1*2)}')\t$(cat ${arr[1]}*/variants/bowtie2/samtools_stats/${arr[0]}.sorted.bam.flagstat | grep '+ 0 mapped' | cut -d ' ' -f1)" | awk '{print $1-($2+$3)}')")\t$(echo "$(echo -e "$(grep 'total_reads' ${arr[1]}*/fastp/${arr[0]}.fastp.json | head -n2 | tail -n1 | cut -d ':' -f2 | sed 's/,//g')\t$(cat ${arr[1]}*/kraken2/${arr[0]}.kraken2.report.txt | grep -v 'unclassified' | cut -f3 | awk '{s+=$1}END{print s}' | awk '{print ($1*2)}')\t$(cat ${arr[1]}*/variants/bowtie2/samtools_stats/${arr[0]}.sorted.bam.flagstat | grep '+ 0 mapped' | cut -d ' ' -f1)" | awk '{print (($1-($2+$3))/$1)*100}')")\t$(cat ${arr[1]}*/multiqc/summary_variants_metrics_mqc.csv | grep ^${arr[0]}, | cut -d ',' -f8)\t$(cat ${arr[1]}*/multiqc/summary_variants_metrics_mqc.csv | grep ^${arr[0]}, | cut -d ',' -f10)\t$(zcat ${arr[1]}*/variants/ivar/consensus/bcftools/${arr[0]}.filtered.vcf.gz | grep -v '^#' | wc -l)\t$(LC_ALL=C awk -F, '{if($10 >= 0.75)print $0}' ${arr[1]}*/variants/ivar/variants_long_table.csv | grep ^${arr[0]}, | grep 'missense' | wc -l)\t$(cat %Ns.tab | grep -w ${arr[0]} | grep ${arr[1]} | cut -f2)\t$(cat */variants/ivar/consensus/bcftools/pangolin/${in}.pangolin.csv | tail -n1 | cut -d ',' -f2)\t$(date '+%Y%m%d')"  >> mapping_illumina_$(date '+%Y%m%d').tab
done
