# Output description for PikaVirus

## Pipeline description
Pipeline for (mainly viral) metagenomic identification using Nextflow.
https://github.com/BU-ISCIII/PikaVirus

## Pipeline overview
* FastQC - v 0.11.9 - Read quality control.
* FastP - v 0.20.1 - Adapter and low quality trimming.
* Kraken 2 - v 2.1.2 - Scouting for each of the
* Krona - v 2.8 - Visualization of kraken's results.
* MASH - v 2.3 - Estimate presence of reference inside the sample reads.
* Bowtie 2 - v 2.4.4 - Map the sample reads against the sample.
* SamTools - v 1.12 - Extract data from the mapping performed by Bowtie 2.
* BedTools - v 2.30.0 - Generate new data from SamTools output.
* MultiQC - v 1.9 - Compile all reports from other programs in one.

## Output directory

- **pikavirus_index.html**: contains a brief list of the params used, as well as a link to every single results sheet created in this run, and a link to the MultiQC report for all samples in the run.

- **`all_samples_{virus, bacteria, fungi}_table.tsv`**: table containing the results for all mappings performed in every sample in the run.

- **multiqc_report.html**: contains a MultiQC report containing the data for all samples in the run (contents will vary according to the set parameters):
    - Preliminar `FastQC` report for **all** samples
    - `FastP` and post-trimming `FastQC` report (only if trimming was set to True) for **all** samples


- **`{sample}_results.html`**: contains the results for each sample in the same page:
    - **Quality control results**:
        - Preliminar `FastQC` report for the sample
        - `FastP` and post-trimming `FastQC` report (only if trimming was set to True) for the sample
        - `MultiQC`report with the above reports
        - Table with the reslts for the control genomes given

    - **Kraken scouting results**:
        - HTML containing the krona plot with all the species identified during the kraken2 step

    - **Coverage results**: for each of the three posible groups chosen (virus, bacteria or fungi):
        - A **general table** containing relevant data of the references found in the sample: *Assembly name*, *Name* (this is, spp of the organism), *Mean depth*, *Standard deviation*, *Minimal depth*, *Maximum depth*, and *% of sequence in or over depths 1, 10, 25, 75 or 100*. <br> The header of the table opens a boxplot with the coverage depth of all references found.  

        - For each of the references in the previous table, a **specific table** containing the same parameters (*Assembly name*, *Mean depth*, *Standard deviation*, *Minimal depth*, *Maximum depth*, and *% of sequence in or over depths 1, 10, 25, 75 or 100* ) is presented. This table contains these entries for the whole genome and for each of the subsequences in it. Also, three different graphs are provided for each entry:
            - **Coverage depth distribtution**: interactive boxplot containing the coverage depth distribution of this subsequence in the sample.

            - **Coverage depth by pos**: interactive plot showing the coverage depth for each position in the reference sequence. (*x axis*: position in the sequence; *y axis*: coverage depth). This plot is available with *max y* = maximum coverage, and *max y* = 500.

            - **Read percentage vs depth**: interactive plot showing the percentage of bases for this reference in this sample, that are over a certain depth. (*x axis*: coverage depth, *y axis*: percentage of bases over that depth).

- **`{sample}_results/`**: directory containing all reports that **`{sample}_results.html`** uses. Please <span style="color: red;">do not delete </span>.

    - **`{sample}_{virus, bacteria, fungi}_table.tsv`**: table containing all mapping results (including phagesm that are excluded by default from results to avoid noise).
    - **fastp.html**: report generated by fastp.
    - **multiqc_report.html**: report generated by MultiQC with all the previous reports.
    - **raw_fastqc/**: directory containing the FastQC result (before trimming, or the only one if no trimming) for this sample.
    - **trimmed_fastqc/**: directory containing the FastQC result (after trimming, will only exist if trimming was performed) for this sample.
    - **`consensus_sequence_{virus, bacteria, fungi}`**: containing the consensus sequences for all references found in the sample, with the name `{organism}_consensus_sequence.fa`. If more than one sequence was found for the same organism, a file containing the MSA for all those consensus will be there, with the name `{organism}_consensus_msa.fa`.
    - **`{virus, bacteria, fungi}_coverage/`**: containing the following:
        - **`{sample}_valid_bedgraph_files_{virus, bacteria, fungi}`**: containing the bedgraph files for all mappings performed.
        - **`{sample}_valid_coverage_files_{virus, bacteria, fungi}`**: containing the coverage files for all mappings performed.
        - **plots**: containing all plots accesible from **`{sample}_results.html`**.
