# ExomeTrio & WGStrio

These file briefly describes the results from ExomeTrio and WGStrio services.

## exomiser.html

This file includes information regarding variant annotation, effect prediction and inheritance typing:

![HTML Description 1](images/exomiser-html-description-1.svg)

![HTML Description 2](images/exomiser-html-description-2.svg)

## picard_hsmetrics.csv

This table includes mapping quality metrics from sarek's pipeline results, with the following columns:
- SAMPLE
- MEAN TARGET COVERAGE: The mean coverage of a target region.
- PCT USABLE BASES ON TARGET: The number of aligned, de-duped, on-bait bases out of the PF bases available (those that pass the vendor's filter).
- FOLD ENRICHMENT: The fold by which the baited region has been amplified above genomic background.
- PCT TARGET BASES 10X: The fraction of all target bases achieving 10X or greater coverage.
- PCT TARGET BASES 20X: The fraction of all target bases achieving 20X or greater coverage.
- PCT TARGET BASES 30X: The fraction of all target bases achieving 30X or greater coverage.
- PCT TARGET BASES 40X: The fraction of all target bases achieving 40X or greater coverage.
- PCT TARGET BASES 50X: The fraction of all target bases achieving 50X or greater coverage.

You may find further documentation for the metrics in this table [here](http://broadinstitute.github.io/picard/picard-metric-definitions.html#HsMetrics).

## variants_annot_filterAF_head.tab

This table includes all the variants from VEP and Exomiser annotation with a minimum Allele Frequency of **0.001**

TO BE COMPLETED

## multiqc_report.html

Most of sarek's QC results are visualised in this report and further statistics are available in the report data directory. 
Results generated by MultiQC collect pipeline QC from supported tools e.g. FastQC. The pipeline has special steps which also allow the software versions to be reported in the MultiQC output for future traceability. For more information about how to use MultiQC reports, see http://multiqc.info.

## Tables in vep_annot folder

These tables are obtained by splitting the variants depending on the mode of inheritance assigned by exomiser. The columns remain the same as the ones found in variants_annot_filterAF.tab
- vep_annot_AD_final.txt: Autosomal Dominant
- vep_annot_AR_final.txt: Autosomal Recessive
- vep_annot_MT_final.txt: Mythocondrial
- vep_annot_XD_final.txt: X-linked Dominant
- vep_annot_XR_final.txt: X-linked Recessive
